crates/bevy_window/src/window.rs

```diff
@@ -145,6 +145,40 @@ impl WindowResizeConstraints {
     }
 }
 
+/// Mirror of `Icon` from `winit`.
+#[derive(Debug, Clone)]
+pub struct WindowIcon {
+    /// 32-bit-per-pixel RGBA image data.
+    pub rgba: Vec<u8>,
+    pub width: u32,
+    pub height: u32,
+}
+
+impl WindowIcon {
+    /// Creates a new `WindowIcon` with possibly unequal width and height.
+    ///
+    /// `rgba` should be 32-bit-per-pixel RGBA image data.
+    #[must_use]
+    pub fn new(rgba: Vec<u8>, width: u32, height: u32) -> Self {
+        Self {
+            rgba,
+            width,
+            height,
+        }
+    }
+
+    /// Creates a new `WindowIcon` with equal width and height.
+    /// `rgba` should be 32-bit-per-pixel RGBA image data.
+    #[must_use]
+    pub fn new_square(rgba: Vec<u8>) -> Self {
```

Comment:
This should warn if it can't be made into a square easily.

Reply:
[e8348b5](https://github.com/bevyengine/bevy/pull/5488/commits/e8348b536d2e554ff69b656709b0a6be5adbff1b) resolves this.